### DISCUSSION TITLE
Dynamically importing icons in react?

### DISCUSSION THREAD

**Discussion by aleyrizvi** [tabler/tabler-icons#379] [Q&A] [answered]
Hi,
I have a menu that is dynamic and the icon name is specified in the api response. 

Is it possible to load icons dynamically in react?

Something like : let Icon = icons[icon] and then <Icon />

Thanks.

**Comment by aleyrizvi** [comment, marked as answer]
For anyone looking for solution, I created a gist for it: https://gist.github.com/aleyrizvi/036b05e948307c28affdb4e16045f2dc

**Reply by dejurin** [reply]
It seems to me that this trick loads all the icons.

**Reply by aleyrizvi** [reply]
Yes, correct.

You can import specific and play with conditional rendering.

**Reply by dejurin** [reply]
Do you have example?

**Reply by aleyrizvi** [reply]
@dejurin something like this:
https://gist.github.com/aleyrizvi/0107975e2a5dbbf11c2ed34db1169eb9

**Reply by dejurin** [reply]
OK

**Comment by sovetski** [comment]
The selected answer is not correct for me, the 1st solution loads all icons which is not optimized, the second one needs that we precise which icons to load which is not logical too because if we need to load the icons dynamically, we can not load them one by one before..

**Reply by evolutionxbox** [reply]
Agreed. Neither of the solutions use the now built-in dynamic imports.

**Comment by ronnyek** [comment]
Is there any new solution for this? I too would like to essentially let someone specify the icon by name, and have it render configured icon dynamically at runtime. 

I'd also like to do this without having to import every possible icon (if possible)

**Reply by sovetski** [reply]
I created an icon picker library and use it personally, I will try to publish it in some days for a public usage

**Reply by aleyrizvi** [reply]
```js
import { useEffect, useState } from "react"

interface Props {
  icon: string
  color?: string
  size?: number
  stroke?: number
}

export function TablerIcon({
  icon,
  color = "gray",
  size = 24,
  stroke = 2,
}: Props) {
  const [Icon, setIcon] = useState<any>(null)

  useEffect(() => {
    let mounted = true

    import(
      /* @vite-ignore */
      `@tabler/icons-react/dist/esm/icons/${icon}`
    ).then((mod) => {
      if (mounted) setIcon(() => mod.default)
    })

    return () => {
      mounted = false
    }
  }, [icon])

  if (!Icon) return null

  return <Icon color={color} size={size} stroke={stroke} />
}
```
